flow: Add docs around clustering (#3572)
grafana/agent: docs/sources/flow/reference/components/prometheus.scrape.md

oauth2 | [oauth2][] | Configure OAuth2 for authenticating to targets. | no
oauth2 > tls_config | [tls_config][] | Configure TLS settings for connecting to targets via OAuth2. | no
tls_config | [tls_config][] | Configure TLS settings for connecting to targets. | no

> clustering | [clustering][] | Configure the component for when the Agent is running in clustered mode | no

Reviewer: Unrelated to this PR, but I was wondering if we have some convention that we use here - seems like we went with a block with a single bool attribute vs. having a bool attribute on the component itself?

Reviewer 2: I suppose we will want to reuse the clustering block for other components in the future, @tpaschalis ?

Author: This could well be a single argument, but we opted it to be a block for two reasons

a) It will be easier to move to a single place to share with other components which implement clustering-specific behavior and add more fields to it.
b) The convention is that each block gets its own little markdown header/section so in this case it allowed us to expand on clustering with some more prose than if we had a single attribute (not that we couldn't do the same under the Attributes section, but it now stands more on its own)

Author: (sorry Paulin, missed your response at first, but yes ^^)

{{< docs/shared lookup="flow/reference/components/tls-config-block.md" source="agent" >}}

### clustering

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`enabled` | `bool` | Enables sharing targets with other cluster nodes.| `false` | yes

When the agent is running in [clustered mode][], and `enabled` is set to true,
then this `prometheus.scrape` component instance opts-in to participating in
the cluster to distribute scrape load between all cluster nodes.

Clustering assumes that all cluster nodes are running with the same
configuration file, have access to the same service discovery APIs and that all
`prometheus.scrape` components that have opted-in to using clustering are
receiving the _exact same_ target set from upstream components in their

> `targets` argument.

Reviewer: Would it be worth describing an impact of frequent changes to discovered targets? There is some race condition there I gather, where different agents would have different targets temporarily.

Author: How about "receiving a consistent target set over their scrape intervals"?

> What happens is that all participating `prometheus.scrape` components use

Reviewer 2: ```suggestion
All `prometheus.scrape` components participating in the cluster use
```

> What happens is that all participating `prometheus.scrape` components use

Reviewer: ```suggestion
When cluster mode is enabled, all participating `prometheus.scrape` components use
```
Nit: Thought this would flow better.

target labels and a consistent hashing algorithm to determine ownership for
each of the targets between the cluster peers. Then, each peer only scrapes the
subset of targets that it is responsible for, so that the scrape load is

> (nearly) evenly split. When a node joins or leaves the cluster, every peer

Reviewer 2: > (nearly) evenly split

Not sure if it worth elaborating more on the "nearly". E.g. why it happens and if it's a problem. Some people will wonder about that as they read the docs.

Author: I'm not sure going into details will help at this phase.

I'm rephrasing this to `so that the scrape load is distributed` and left it at that for now.

re-calculates ownership and continues scraping with the new target set. This

> performs better than hashmod sharding where _all_ nodes have to be

Reviewer 2: Worth noting drawbacks compared to hashmod? E.g. I suppose eventual consistency vs full consistency?

Author: Yes, that's the main argument. Adding it as a note here.

{{< docs/shared lookup="flow/reference/components/tls-config-block.md" source="agent" >}}

### clustering (experimental)

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`enabled` | `bool` | Enables sharing targets with other cluster nodes. | `false` | yes

When the agent is running in [clustered mode][], and `enabled` is set to true,
then this `prometheus.scrape` component instance opts-in to participating in
the cluster to distribute scrape load between all cluster nodes.

Clustering assumes that all cluster nodes are running with the same
configuration file, have access to the same service discovery APIs and that all
`prometheus.scrape` components that have opted-in to using clustering are
receiving the _exact same_ target set from upstream components in their
`targets` argument.

All `prometheus.scrape` components instances opting in to clustering use target
labels and a consistent hashing algorithm to determine ownership for each of
the targets between the cluster peers. Then, each peer only scrapes the subset
of targets that it is responsible for, so that the scrape load is distributed. 

> When a node joins or leaves the cluster, every peer re-calculates ownership and

Reviewer 2: ```suggestion
When a node joins or leaves the cluster, every peer recalculates ownership and
```

Author: Thanks, done.

continues scraping with the new target set. This performs better than hashmod
sharding where _all_ nodes have to be re-distributed, as only 1/N of the
targets ownership is transferred, but is eventually consistent (rather than

> fully consistent like hashmod sharding is).

Reviewer 3: Link to hashmod sharding example?

Author: I don't think we have a good Flow example right now for this, but let me look.
